driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/syncadapter/SyncMongoCursor.java

```diff
@@ -16,37 +16,138 @@
 
 package com.mongodb.reactivestreams.client.syncadapter;
 
+import com.mongodb.MongoInterruptedException;
+import com.mongodb.MongoTimeoutException;
 import com.mongodb.ServerAddress;
 import com.mongodb.ServerCursor;
 import com.mongodb.client.MongoCursor;
 import com.mongodb.lang.Nullable;
 import org.reactivestreams.Publisher;
-import reactor.core.publisher.Flux;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
 
-import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.ClusterFixture.TIMEOUT;
 
 class SyncMongoCursor<T> implements MongoCursor<T> {
-    private final Iterator<T> iterator;
+    private static final Object COMPLETED = new Object();
+    private final BlockingDeque<Object> results = new LinkedBlockingDeque<>();
+    private final Integer batchSize;
+    private int countToBatchSize;
+    private volatile Subscription subscription;
+    private T current;
+    private boolean completed;
+    private RuntimeException error;
+
+    SyncMongoCursor(final Publisher<T> publisher, @Nullable final Integer batchSize) {
```

Comment:
Because we have new assertions in the unified test runner.  For example, there is a test which assert that a killCursors command is sent.  The waiting in close method gives a chance for the asynchronous close action to do that.

Reply:
Discussed this a bit more via Slack, and https://jira.mongodb.org/browse/JAVA-4173 is a result of that discussion.